packages/aws-cdk/lib/api/hotswap/monitor/logs-monitor.ts

```diff
@@ -0,0 +1,226 @@
+  /**
+   * Time that the latest hotswap operation occurred
+   *
+   * @default - local machine's current time
+   */
+  readonly hotswapTime?: Date;
+
+}
+
+export interface CloudWatchLogEvent {
+  readonly message: string;
+  readonly logGroup: string;
+  readonly logStream?: string;
+  readonly ingestionTime: Date;
+}
+
+export class CloudWatchLogEventMonitor {
+  public static withDefaultPrinter(options: WithDefaultPrinterProps = {}) {
+    const stream = process.stderr;
+    const props: PrinterProps = {
+      stream,
+    };
+
+    return new CloudWatchLogEventMonitor(new EventPrinter(props), options.hotswapTime);
+  }
+
+  private active = false;
+
+  /**
+   * Determines which events not to display
+   */
+  private startTime: number;
+
+  /**
+   * Current tick timer
+   */
+  private tickTimer?: NodeJS.Timer;
+
+  /**
+   * Set to the activity of reading the current events
+   */
+  private readPromise?: Promise<any>;
+
+  private activity = new Set<string>();
+
+  private logGroups = new Set<string>();
+  private sdk?: ISDK;
+
+  constructor(
+    private readonly printer: IEventPrinter,
+    hotswapTime?: Date,
+  ) {
+    this.startTime = hotswapTime?.getTime() ?? Date.now();
+  }
+
+  public start() {
+    this.active = true;
+    this.scheduleNextTick();
+    return this;
+  }
+
+  public addSDK(sdk: ISDK) {
+    this.sdk = sdk;
+  }
+
+  /**
+   * Adds a CloudWatch log group to read log events from
+   */
+  public addLogGroups(logGroups: string[]) {
+    logGroups.forEach(group => {
+      if (!this.logGroups.has(group)) {
+        this.logGroups.add(group);
+      }
+    });
+  }
+
+  public async stop() {
+    this.active = false;
+    if (this.tickTimer) {
+      clearTimeout(this.tickTimer);
+    }
+  }
+
+  private scheduleNextTick() {
+    if (!this.active) {
+      return;
+    }
+    this.tickTimer = setTimeout(() => void (this.tick()), this.printer.updateSleep);
+  }
+
+  private async tick() {
+    if (!this.active) {
+      return;
+    }
+    try {
+      this.readPromise = this.readNewEvents();
+      await this.readPromise;
+      this.readPromise = undefined;
+
+      if (!this.active) { return; }
+
+    } catch (e) {
+      error('Error occurred while monitoring logs: %s', e);
+    }
+    this.scheduleNextTick();
+
+  }
+
+  /**
+   * Reads all new log events from a set of CloudWatch Log Groups
+   * in parallel
+   */
+  private async readNewEvents(): Promise<void[]> {
+    const promises: Array<Promise<void>> = [];
+    for (let group of this.logGroups) {
+      promises.push(this.readEventsFromLogGroup(group));
+    }
+    return Promise.all(promises);
+  }
+
+  /**
+   * Reads all new log events from a CloudWatch Log Group
+   * starting when the hotswap was triggered
+   *
+   * Only prints out events that have not been printed already
+   */
+  private async readEventsFromLogGroup(logGroupName: string): Promise<void> {
+    // this should not be possible
+    if (!this.sdk) {
+      return;
+    }
```

Comment:
Let's get rid of this check then, especially that you use `this.sdk!` below anyway.